File: Assignment18/Assignment18/score_processor.py
"""
ScoreProcessor module for reading and processing score data from files.
Handles file I/O with proper exception handling and resource cleanup.
"""


class ScoreProcessor:
    """
    Processes score data from files with comprehensive error handling.
    
    Reads an integer score from a file, multiplies it by 10, and returns the result.
    Handles file not found and invalid format errors gracefully.
    Ensures proper resource cleanup in all scenarios.
    """
    
    # Multiplication factor for score calculation
    SCORE_MULTIPLIER = 10
    
    def process_score_file(self, file_path: str) -> int:
        """
        Process a score from a file by reading an integer value,
        multiplying it by 10, and returning the result.
        
        Handles file not found and invalid number format errors gracefully.
        Ensures proper resource cleanup in all scenarios using try-except-else-finally.
        
        Args:
            file_path: The path to the file containing the score
            
        Returns:
            The score multiplied by 10
            
        Raises:
            FileNotFoundError: If the file does not exist
            ValueError: If the file content is not a valid integer
        """
        file_handle = None
        
        try:
            # Open the file
            file_handle = open(file_path, 'r')
            
            # Read the first line
            line = file_handle.readline()
            
            # Check if file is empty
            if not line or not line.strip():
                raise ValueError("File is empty or contains no data")
            
            # Parse the line as an integer
            score = int(line.strip())
            
            # Calculate result
            result = score * self.SCORE_MULTIPLIER
            
            print(f"Successfully read score: {score}")
            print(f"Calculated result: {result}")
            
            return result
            
        except FileNotFoundError as e:
            # Handle missing file
            print(f"ERROR: File not found at path: {file_path}")
            print(f"Details: {e}")
            raise
            
        except ValueError as e:
            # Handle invalid number format
            print("ERROR: Invalid number format in file")
            print(f"Details: {e}")
            print("Expected: A single integer value")
            raise
            
        except IOError as e:
            # Handle other I/O errors
            print("ERROR: I/O error while reading file")
            print(f"Details: {e}")
            raise RuntimeError(f"Failed to read file: {e}") from e
            
        else:
            # Executed if no exception occurred
            print("Data processed successfully")
            
        finally:
            # Cleanup - always executed
            if file_handle is not None:
                try:
                    file_handle.close()
                    print("File cleanup completed")
                except IOError as e:
                    print(f"ERROR: Failed to close file: {e}")
    
    def process_score_file_with_context(self, file_path: str) -> int:
        """
        Alternative implementation using context manager (with statement).
        This is the Pythonic way to handle file I/O.
        
        Args:
            file_path: The path to the file containing the score
            
        Returns:
            The score multiplied by 10
            
        Raises:
            FileNotFoundError: If the file does not exist
            ValueError: If the file content is not a valid integer
        """
        try:
            # Context manager automatically closes the file
            with open(file_path, 'r') as file_handle:
                # Read the first line
                line = file_handle.readline()
                
                # Check if file is empty
                if not line or not line.strip():
                    raise ValueError("File is empty or contains no data")
                
                # Parse the line as an integer
                score = int(line.strip())
                
                # Calculate result
                result = score * self.SCORE_MULTIPLIER
                
                print(f"Successfully read score: {score}")
                print(f"Calculated result: {result}")
                
                return result
                
        except FileNotFoundError as e:
            # Handle missing file
            print(f"ERROR: File not found at path: {file_path}")
            print(f"Details: {e}")
            raise
            
        except ValueError as e:
            # Handle invalid number format
            print("ERROR: Invalid number format in file")
            print(f"Details: {e}")
            print("Expected: A single integer value")
            raise
            
        except IOError as e:
            # Handle other I/O errors
            print("ERROR: I/O error while reading file")
            print(f"Details: {e}")
            raise RuntimeError(f"Failed to read file: {e}") from e
            
        else:
            # Executed if no exception occurred
            print("Data processed successfully")
            
        finally:
            # Cleanup message - always executed
            print("File cleanup completed")
    
    def get_score_multiplier(self) -> int:
        """
        Get the score multiplier constant.
        
        Returns:
            The multiplier value (10)
        """
        return self.SCORE_MULTIPLIER

File: Assignment18/ScoreProcessor.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileNotFoundException;
import java.io.IOException;

/**
 * ScoreProcessor class for reading and processing score data from files.
 * Handles file I/O with proper exception handling and resource cleanup.
 */
public class ScoreProcessor {
    
    // Multiplication factor for score calculation
    private static final int SCORE_MULTIPLIER = 10;
    
    /**
     * Processes a score from a file by reading an integer value,
     * multiplying it by 10, and returning the result.
     * 
     * Handles file not found and invalid number format errors gracefully.
     * Ensures proper resource cleanup in all scenarios.
     * 
     * @param filePath the path to the file containing the score
     * @return the score multiplied by 10
     * @throws FileNotFoundException if the file does not exist
     * @throws NumberFormatException if the file content is not a valid integer
     */
    public int processScoreFile(String filePath) throws FileNotFoundException {
        int score = 0;
        
        try {
            // Try-with-resources automatically closes the BufferedReader
            try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
                String line = reader.readLine();
                
                // Check if file is empty
                if (line == null || line.trim().isEmpty()) {
                    throw new NumberFormatException("File is empty or contains no data");
                }
                
                // Parse the line as an integer
                score = Integer.parseInt(line.trim());
                
                // Calculate result
                int result = score * SCORE_MULTIPLIER;
                
                System.out.println("Successfully read score: " + score);
                System.out.println("Calculated result: " + result);
                
                return result;
            }
            
        } catch (FileNotFoundException e) {
            // Handle missing file
            System.err.println("ERROR: File not found at path: " + filePath);
            System.err.println("Details: " + e.getMessage());
            throw e;
            
        } catch (NumberFormatException e) {
            // Handle invalid number format
            System.err.println("ERROR: Invalid number format in file");
            System.err.println("Details: " + e.getMessage());
            System.err.println("Expected: A single integer value");
            throw e;
            
        } catch (IOException e) {
            // Handle other I/O errors
            System.err.println("ERROR: I/O error while reading file");
            System.err.println("Details: " + e.getMessage());
            throw new RuntimeException("Failed to read file: " + e.getMessage(), e);
            
        } finally {
            // Cleanup message - always executed
            System.out.println("File cleanup completed");
        }
    }
    
    /**
     * Alternative implementation using explicit try-catch-finally
     * (without try-with-resources) for demonstration purposes.
     * 
     * @param filePath the path to the file containing the score
     * @return the score multiplied by 10
     */
    public int processScoreFileExplicit(String filePath) throws FileNotFoundException {
        BufferedReader reader = null;
        
        try {
            // Open the file
            reader = new BufferedReader(new FileReader(filePath));
            String line = reader.readLine();
            
            // Check if file is empty
            if (line == null || line.trim().isEmpty()) {
                throw new NumberFormatException("File is empty or contains no data");
            }
            
            // Parse the line as an integer
            int score = Integer.parseInt(line.trim());
            
            // Calculate result
            int result = score * SCORE_MULTIPLIER;
            
            System.out.println("Successfully read score: " + score);
            System.out.println("Calculated result: " + result);
            
            return result;
            
        } catch (FileNotFoundException e) {
            // Handle missing file
            System.err.println("ERROR: File not found at path: " + filePath);
            System.err.println("Details: " + e.getMessage());
            throw e;
            
        } catch (NumberFormatException e) {
            // Handle invalid number format
            System.err.println("ERROR: Invalid number format in file");
            System.err.println("Details: " + e.getMessage());
            System.err.println("Expected: A single integer value");
            throw e;
            
        } catch (IOException e) {
            // Handle other I/O errors
            System.err.println("ERROR: I/O error while reading file");
            System.err.println("Details: " + e.getMessage());
            throw new RuntimeException("Failed to read file: " + e.getMessage(), e);
            
        } finally {
            // Cleanup - close the reader if it was opened
            if (reader != null) {
                try {
                    reader.close();
                    System.out.println("File cleanup completed");
                } catch (IOException e) {
                    System.err.println("ERROR: Failed to close file: " + e.getMessage());
                }
            }
        }
    }
    
    /**
     * Gets the score multiplier constant.
     * 
     * @return the multiplier value (10)
     */
    public int getScoreMultiplier() {
        return SCORE_MULTIPLIER;
    }
}
